### PR TITLE
Implement Cilium Ingress

### LINKED ISSUE #15784: Add Cilium Ingress suport
/kind feature

**1. Describe IN DETAIL the feature/behavior/change you would like to see.**

We have now the ability to add cilium ingress support: https://docs.cilium.io/en/v1.13/network/servicemesh/ingress/
It allows users to use the cilium ingress controller instead of using nginx, haproxy etc.

As far as I know, this will also allow to use L7 proxy directly with Cilium.
Especially with Cilium 1.14, which can handle Wireguard + L7 Proxy, this becomes more interesting.

**2. Feel free to provide a design supporting your feature request.**

I'll try to provide a PR in the next few days. It should be implemented as we have done with Hubble (via CiliumIngressSpec in CiliumNetworkingSpec).

### PR DESCRIPTION
**Pull request by zadjadr**
Fixes #15784

### CODE REVIEW

File: pkg/apis/kops/validation/validation.go [hunk 1]
```diff
@@ -1353,6 +1353,12 @@ func validateNetworkingCilium(cluster *kops.Cluster, v *kops.CiliumNetworkingSpe
 		}
 	}
 
+	if v.Ingress != nil && fi.ValueOf(v.Ingress.Enabled) {
+		if v.Ingress.DefaultLoadBalancerMode != "" {
+			allErrs = append(allErrs, IsValidValue(fldPath.Child("ingress", "defaultLoadBalancerMode"), &v.Ingress.DefaultLoadBalancerMode, []string{"shared", "dedicated"})...)
+		}
+	}
```
Comment (olemarkus) [inline pkg/apis/kops/validation/validation.go:1360]: Is it allowed to set SharedLoadBalancerServiceName if the mode is dedicated?

File: pkg/model/components/cilium.go [hunk 2]
```diff
@@ -168,5 +168,34 @@ func (b *CiliumOptionsBuilder) BuildOptions(o interface{}) error {
...
+			ingress.EnableSecretsSync = fi.PtrTo(true)
+		}
+		if ingress.SecretsNamespace == "" {
+			ingress.SecretsNamespace = "cilium-secrets"
+		}
+		if ingress.LoadBalancerAnnotationPrefixes == "" {
+			ingress.LoadBalancerAnnotationPrefixes = "service.beta.kubernetes.io service.kubernetes.io cloud.google.com"
+		}
+		if ingress.DefaultLoadBalancerMode == "" {
+			ingress.DefaultLoadBalancerMode = "shared"
```
Comment (olemarkus) [inline pkg/model/components/cilium.go:189]: Should shared really be the default? Cilium has dedicated as default.

File: pkg/apis/kops/v1alpha3/networking.go [hunk 3]
```diff
@@ -456,6 +456,40 @@ type CiliumNetworkingSpec struct {
...
+
+	// EnforceHttps specifies whether HTTPS enforcement is enabled for Ingress traffic.
+	// Default: true
+	EnforceHttps *bool `json:"enforceHttps,omitempty"`
+
+	// EnableSecretsSync specifies whether synchronization of secrets is enabled.
+	// Default: true
+	EnableSecretsSync *bool `json:"enableSecretsSync,omitempty"`
+
+	// SecretsNamespace specifies the namespace where secrets are synchronized.
```
Comment (olemarkus) [inline pkg/apis/kops/v1alpha3/networking.go:477]: I am not sure we need to make this one configurable.

File: pkg/model/components/cilium.go [hunk 4]
```diff
@@ -168,5 +168,34 @@ func (b *CiliumOptionsBuilder) BuildOptions(o interface{}) error {
...
+			ingress.Enabled = fi.PtrTo(true)
+		}
+		if ingress.EnforceHttps == nil {
+			ingress.EnforceHttps = fi.PtrTo(true)
+		}
+		if ingress.EnableSecretsSync == nil {
+			ingress.EnableSecretsSync = fi.PtrTo(true)
+		}
+		if ingress.SecretsNamespace == "" {
+			ingress.SecretsNamespace = "cilium-secrets"
```
Comment (olemarkus) [inline pkg/model/components/cilium.go:183]: We tend to reuse `kube-system` for things like this as it's the one namespace kOps owns.

**Review by olemarkus — CHANGES_REQUESTED** [review]

File: pkg/model/components/cilium.go (hunk 4 above)
Comment (zadjadr) [reply pkg/model/components/cilium.go:183]: Sure, we can use `kube-system` - I just used the default namespace provided in the [cilium helm chart](https://github.com/cilium/cilium/blob/v1.13.6/install/kubernetes/cilium/values.yaml#L616). I'd just go the kOps way then :+1:

File: pkg/model/components/cilium.go (hunk 2 above)
Comment (zadjadr) [reply pkg/model/components/cilium.go:189]: You're right makes more sense to use dedicated if cilium uses that as well

File: pkg/apis/kops/validation/validation.go (hunk 1 above)
Comment (zadjadr) [reply pkg/apis/kops/validation/validation.go:1360]: Might make sense to completely remove this option. The cilium helm chart does not make the name configurable. Wdyt?

Comment (zadjadr) [reply pkg/apis/kops/validation/validation.go:1360]: At least from the helm chart, I see that it is always set and does not depend on the mode:

https://github.com/cilium/cilium/blob/main/install/kubernetes/cilium/templates/cilium-configmap.yaml#L248

**Comment by zadjadr** [comment]
@olemarkus I have added all your suggestions here (except for the possible check on [shared lb name](https://github.com/kubernetes/kops/pull/15795#discussion_r1300027488)): https://github.com/kubernetes/kops/pull/15795/commits/16a7b9bf6381ee3466744c6782725518f81b1ce5

I'll squash the commits if your happy with the changes.

File: pkg/model/components/cilium.go [hunk 5]
```diff
@@ -168,5 +168,31 @@ func (b *CiliumOptionsBuilder) BuildOptions(o interface{}) error {
...
+			ingress.DefaultLoadBalancerMode = "dedicated"
+		}
+		if ingress.SharedLoadBalancerServiceName == "" {
+			ingress.SharedLoadBalancerServiceName = "cilium-ingress"
+		}
+	} else {
+		c.Ingress = &kops.CiliumIngressSpec{
+			Enabled: fi.PtrTo(false),
+		}
+	}
```
Comment (hakman) [inline pkg/model/components/cilium.go:195]: Can we try something like this, to not populate all the parameters if ingress is not enabled?
```suggestion
	if ingress == nil {
		c.Ingress = &kops.CiliumIngressSpec{
			Enabled: fi.PtrTo(false),
		}
	}
	if ingress.Enabled == nil {
		ingress.Enabled = fi.PtrTo(true)
	}
	if fi.ValueOf(ingress.Enabled ) {
		if ingress.EnforceHttps == nil {
			ingress.EnforceHttps = fi.PtrTo(true)
		}
		if ingress.EnableSecretsSync == nil {
			ingress.EnableSecretsSync = fi.PtrTo(true)
		}
		if ingress.LoadBalancerAnnotationPrefixes == "" {
			ingress.LoadBalancerAnnotationPrefixes = "service.beta.kubernetes.io service.kubernetes.io cloud.google.com"
		}
		if ingress.DefaultLoadBalancerMode == "" {
			ingress.DefaultLoadBalancerMode = "dedicated"
		}
		if ingress.SharedLoadBalancerServiceName == "" {
			ingress.SharedLoadBalancerServiceName = "cilium-ingress"
		}
	}
```

**Review by hakman — CHANGES_REQUESTED** [review]

File: pkg/model/components/cilium.go (hunk 5 above)
Comment (zadjadr) [reply pkg/model/components/cilium.go]: Looks good :) I'll clean up the commits in the end

File: pkg/model/components/cilium.go [hunk 6]
```diff
@@ -168,5 +168,32 @@ func (b *CiliumOptionsBuilder) BuildOptions(o interface{}) error {
...
+	ingress := c.Ingress
+	if ingress == nil {
+		c.Ingress = &kops.CiliumIngressSpec{
+			Enabled: fi.PtrTo(false),
+		}
+	}
+	if ingress.Enabled == nil {
+		ingress.Enabled = fi.PtrTo(true)
+	}
+	if fi.ValueOf(ingress.Enabled ) {
```
Comment (hakman) [inline pkg/model/components/cilium.go:180]:
```suggestion
	if fi.ValueOf(ingress.Enabled) {
```

File: pkg/model/components/cilium.go (hunk 5 above)
Comment (zadjadr) [reply pkg/model/components/cilium.go]: @hakman looks like this results in nil pointer dereference :(

https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/kops/15795/pull-kops-verify-golangci-lint/1694393521546465280

Any idea how to fix this?

File: pkg/model/components/cilium.go [hunk 7]
```diff
@@ -168,5 +168,32 @@ func (b *CiliumOptionsBuilder) BuildOptions(o interface{}) error {
...
+		if ingress.LoadBalancerAnnotationPrefixes == "" {
+			ingress.LoadBalancerAnnotationPrefixes = "service.beta.kubernetes.io service.kubernetes.io cloud.google.com"
+		}
+		if ingress.DefaultLoadBalancerMode == "" {
+			ingress.DefaultLoadBalancerMode = "dedicated"
+		}
+		if ingress.SharedLoadBalancerServiceName == "" {
+			ingress.SharedLoadBalancerServiceName = "cilium-ingress"
+		}
+	}
```
Comment (hakman) [inline pkg/model/components/cilium.go:196]:
```suggestion
	if c.Ingress == nil {
		c.Ingress = &kops.CiliumIngressSpec{
			Enabled: fi.PtrTo(false),
		}
	}
	if c.Ingress.Enabled == nil {
		c.Ingress.Enabled = fi.PtrTo(true)
	}
	if fi.ValueOf(c.Ingress.Enabled) {
		if c.Ingress.EnforceHttps == nil {
			c.Ingress.EnforceHttps = fi.PtrTo(true)
		}
		if c.Ingress.EnableSecretsSync == nil {
			c.Ingress.EnableSecretsSync = fi.PtrTo(true)
		}
		if c.Ingress.LoadBalancerAnnotationPrefixes == "" {
			c.Ingress.LoadBalancerAnnotationPrefixes = "service.beta.kubernetes.io service.kubernetes.io cloud.google.com"
		}
		if c.Ingress.DefaultLoadBalancerMode == "" {
			c.Ingress.DefaultLoadBalancerMode = "dedicated"
		}
		if c.Ingress.SharedLoadBalancerServiceName == "" {
			c.Ingress.SharedLoadBalancerServiceName = "cilium-ingress"
		}
	}
```

File: pkg/model/components/cilium.go (hunk 5 above)
Comment (hakman) [reply pkg/model/components/cilium.go]: Yes, see my suggestion. You will also need to run hack/update-expected.

**Review by olemarkus — CHANGES_REQUESTED** [review]

File: pkg/model/components/cilium.go [hunk 8]
```diff
@@ -168,5 +168,31 @@ func (b *CiliumOptionsBuilder) BuildOptions(o interface{}) error {
...
+	if fi.ValueOf(c.Ingress.Enabled) {
+		if c.Ingress.EnforceHttps == nil {
+			c.Ingress.EnforceHttps = fi.PtrTo(true)
+		}
+		if c.Ingress.EnableSecretsSync == nil {
+			c.Ingress.EnableSecretsSync = fi.PtrTo(true)
+		}
+		if c.Ingress.LoadBalancerAnnotationPrefixes == "" {
+			c.Ingress.LoadBalancerAnnotationPrefixes = "service.beta.kubernetes.io service.kubernetes.io cloud.google.com"
```
[the remark this answers is not in the document]
Comment (zadjadr) [reply pkg/model/components/cilium.go:187]: Where can I read up on that? I know the helm chart sets a default value in the configmap, but no idea if cilium will set it if not set in the configmap

File: upup/models/cloudup/resources/addons/networking.cilium.io/k8s-1.16-v1.13.yaml.template [hunk 9]
```diff
@@ -609,6 +638,119 @@ subjects:
...
+roleRef:
+  apiGroup: rbac.authorization.k8s.io
+  kind: Role
+  name: cilium-operator-ingress-secrets
+subjects:
+  - kind: ServiceAccount
+    name: "cilium-operator"
+    namespace: kube-system
+---
+# Source: cilium/templates/cilium-ingress-service.yaml
```
[the remark this answers is not in the document]
Comment (zadjadr) [reply upup/models/cloudup/resources/addons/networking.cilium.io/k8s-1.16-v1.13.yaml.template:714]: Atleast the helm chart always creates these if ingress is enabled: https://github.com/cilium/cilium/blob/v1.14.1/install/kubernetes/cilium/templates/cilium-ingress-service.yaml

But it makes sense to disable this behavior in kops. Should we go against the helm chart here?

By the way, just curious about this: why do we not create a "values.yaml" and generate the entire cilium `k8s-1.16-v1.13.yaml.template` via helms go library?

File: pkg/model/components/cilium.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (zadjadr) [reply pkg/model/components/cilium.go:187]: Okay found the default options here: https://pkg.go.dev/github.com/cilium/cilium/operator/option

I'll remove the stuff that can be left blank.

File: pkg/model/components/cilium.go [hunk 10]
```diff
@@ -168,5 +168,31 @@ func (b *CiliumOptionsBuilder) BuildOptions(o interface{}) error {
...
+		}
+		if c.Ingress.DefaultLoadBalancerMode == "" {
+			c.Ingress.DefaultLoadBalancerMode = "dedicated"
+		}
+		if c.Ingress.SharedLoadBalancerServiceName == "" {
+			c.Ingress.SharedLoadBalancerServiceName = "cilium-ingress"
```
[the remark this answers is not in the document]
Comment (zadjadr) [reply pkg/model/components/cilium.go:193]: Where did you find the default value set by cilium? I can't find it anywhere in https://pkg.go.dev/github.com/cilium/cilium/operator/option or in https://github.com/cilium/cilium/tree/main/pkg/defaults

File: upup/models/cloudup/resources/addons/networking.cilium.io/k8s-1.16-v1.13.yaml.template (hunk 9 above)
[the remark this answers is not in the document]
Comment (hakman) [reply upup/models/cloudup/resources/addons/networking.cilium.io/k8s-1.16-v1.13.yaml.template:714]: That may be an interesting approach for the future.

File: pkg/model/components/cilium.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (hakman) [reply pkg/model/components/cilium.go:193]: I think these are the ones: https://github.com/cilium/cilium/blob/main/operator/pkg/ingress/ingress_options.go#L21-L31.

**Review by hakman — APPROVED** [review]
Thanks @zadjadr, LGTM!

**Pull request merged**